plugins/chain_plugin/chain_plugin.cpp

```diff
@@ -797,35 +822,97 @@ void chain_plugin::plugin_initialize(const variables_map& options) {
+            const chain_id_type passed_chain_id = get_chain_id();
+            if( my->chain_id ) {
+               ilog("REMOVE chain id: ${ci}, passed_chain_id: ${gci}", ("ci", *my->chain_id)("gci",passed_chain_id));
+               EOS_ASSERT( *my->chain_id == passed_chain_id, plugin_config_exception,
+                           "chain id provided via command line arguments does not match the existing chain id in the ${source}. "
+                           "It is not necessary to provide a chain id argument when loading a snapshot.", ("source", my->chain_id_source)
+               );
+            }
+            else {
+               my->chain_id = passed_chain_id;
+               my->chain_id_source = "chain id command line parameter";
+            }
+         }
+
          auto shared_mem_path = my->chain_config->state_dir / "shared_memory.bin";
          EOS_ASSERT( !fc::exists(shared_mem_path),
                  plugin_config_exception,
                  "Snapshot can only be used to initialize an empty database." );
 
          if( fc::is_regular_file( my->blocks_dir / "blocks.log" )) {
-            auto log_genesis = block_log::extract_genesis_state(my->blocks_dir);
-            EOS_ASSERT( log_genesis.compute_chain_id() == my->chain_config->genesis.compute_chain_id(),
-                    plugin_config_exception,
-                    "Genesis information in blocks.log does not match genesis information in the snapshot");
+            const auto log_chain_id = block_log::extract_chain_id(my->blocks_dir);
+            if( my->chain_id ) {
+               EOS_ASSERT( log_chain_id == *my->chain_id,
+                           plugin_config_exception,
+                           "Chain id in blocks.log does not match chain id in the snapshot");
+            }
+            else {
+               my->chain_id = log_chain_id;
+               my->chain_id_source = "block log";
+            }
          }
 
+         if (my->chain_id) {
+            ilog("chain id set to: ${ci} by ${source}", ("ci", my->chain_id)("source", my->chain_id_source));
+         }
+         if (my->chain_config->genesis) {
+            ilog("genesis set to: ${ci}", ("ci", *my->chain_config->genesis));
+         }
       } else {
          bfs::path genesis_file;
          bool genesis_timestamp_specified = false;
          fc::optional<genesis_state> existing_genesis;
 
          if( fc::exists( my->blocks_dir / "blocks.log" ) ) {
-            my->chain_config->genesis = block_log::extract_genesis_state( my->blocks_dir );
-            existing_genesis = my->chain_config->genesis;
+            existing_genesis = block_log::extract_genesis_state( my->blocks_dir );
+            EOS_ASSERT( existing_genesis,
+                        plugin_config_exception,
+                        "Specified genesis file '${genesis}' does not contain a genesis state (truncated block log).  "
+                        "Truncated block logs can only be used with snapshots.",
```

Comment:
This is not true. Truncated blocks logs are perfectly fine to use without the `--snapshot` option.

This whole `else` branch (i.e. when not using `--snapshot`) can also be greatly simplified. Here is the basic idea:

Within `plugin_initialize`:
1. If a `--genesis-json` is provided: use it (and `--genesis-timestamp` if also provided, but do not allow `--genesis-timestamp` to be provided without also providing `--genesis-json`) to compute the `genesis_state` effectively provided by the user; store it in a `optional<genesis_state>` member variable called `provided_genesis_state` within `chain_plugin_impl`; and then store `provided_genesis_state->compute_chain_id()` in `expected_chain_id` within `chain_plugin_impl`.
2. If a blocks.log exists: determine whether a genesis state can be extracted from it. If so, extract the genesis state, compute the chain ID from it, ensure that the computed chain ID matches one in `expected_chain_id` (if present), store the chain ID in `expected_chain_id` (if it was not already present), and store the extract genesis state (if it existed in the blocks.log) in `provided_genesis_state`.
3. At this point, `expected_chain_id` may or may not have a value and `provided_genesis_state` may or may not have a value. But if `provided_genesis_state` has a value then so will `expected_chain_id`.

Then later in `plugin_startup`, the following is done:

```cpp
void chain_plugin::plugin_startup()
{ try {
   try {
      auto shutdown = [](){ return app().is_quiting(); };
      if( my->snapshot_path ) {
         auto infile = std::ifstream(my->snapshot_path->generic_string(), (std::ios::in | std::ios::binary));
         auto reader = std::make_shared<istream_snapshot_reader>(infile);
         my->chain->startup( shutdown, reader );
         infile.close();
      } else if( my->chain->db().revision() < 1 ) {
         EOS_ASSERT( my->provided_genesis_state || !my->expected_chain_id, plugin_config_exception,
                     "Genesis state is necessary to initialize fresh blockchain state but genesis state could not be found in the blocks log. Please either load from snapshot or find a blocks log that starts from genesis."
         );
         if( my->provided_genesis_state ) {
            ilog( "Starting up fresh blockchain with provided genesis state." );
         } else {
            ilog( "Starting up fresh blockchain with default genesis state." );
         }
         my->chain->startup( shutdown, (my->provided_genesis_state ? *my->provided_genesis_state : genesis_state{}) );
      } else {
         if( my->expected_chain_id && (*my->expected_chain_id != my->chain->db().get<global_property_object>().chain_id) ) {
            EOS_ASSERT( my->provided_genesis_state, block_log_exception, 
                        "Existing state is for a different chain than the one specified in the blocks log.",
                        ("blocks_log_chain_id", *my->expected_chain_id)
                        ("state_chain_id", my->chain->db().get<global_property_object>().chain_id)
                        
            );
            EOS_THROW( plugin_config_exception, 
                       "Existing state is for a different chain than the one specified by the provided genesis state.",
                       ("provided_chain_id", *my->expected_chain_id)
                       ("state_chain_id", my->chain->db().get<global_property_object>().chain_id)
            );
         }
         my->chain->startup( shutdown );
      }
   } catch (const database_guard_exception& e) {
// ...
```